### PR TITLE
AVRO-3985: Add trusted packages support in SpecificData

### PR DESCRIPTION
**Pull request by jbonofre**
## What is the purpose of the change

This change introduces the `org.apache.avro.SERIALIZABLE_PACKAGES` system property to enforce the security aspect of using `java-class` in a schema.

## Verifying this change

This change is already covered by existing tests, using the default trusted packages. I can add an additional test specifically for non trusted packages.

## Documentation

- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? The error message already explains how to use `org.apache.avro.SERIALIZABLE_PACKAGES` system property. Happy to add additional documentation if needed.

### CODE REVIEW

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java [hunk 1]
```diff
@@ -372,6 +372,7 @@ private static boolean isBlank(CharSequence cs) {
 
   /** Return the class that implements a schema, or null if none exists. */
   public Class getClass(Schema schema) {
+
```
Comment (Fokko) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificData.java:375]: Let's avoid unrelated changes:
```suggestion
```

**Review by Fokko — APPROVED** [review]
Thanks @jbonofre

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflectData.java [hunk 2]
```diff
@@ -40,6 +41,17 @@
...
 public class TestReflectData {
+
+  @Test
+  void trustedPackages() throws Exception {
+    Schema schema = new Schema.Parser().parse(new File("target/test-classes/schema-trusted-packages.json"));
+
+    Schema.Field field = schema.getField("value");
+    System.out.println(field);
+
+    System.out.println(schema);
```
Comment (martin-g) [inline lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflectData.java:52]: This test looks like a playground. 
It should be either finished (e.g. by using assertions) or removed.

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java [hunk 3]
```diff
@@ -1171,6 +1171,7 @@ public static class NullableStringable {
 
   @Test
   void nullableStringableField() throws Exception {
+    System.setProperty("org.apache.avro.TRUSTED_PACKAGES", "*");
```
[the remark this answers is not in the document]
Comment (jbonofre) [reply lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflect.java:1174]: Actually i think we don't need it with the default trusted packages. Let me double check. Thanks !

File: lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflectData.java (hunk 2 above)
Comment (jbonofre) [reply lang/java/avro/src/test/java/org/apache/avro/reflect/TestReflectData.java:52]: It should be removed (my bad sorry about that). I will clean up.

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 4]
```diff
@@ -24,12 +24,26 @@
...
+
+  public static final String[] SERIALIZABLE_PACKAGES;
+
+  static {
+    SERIALIZABLE_PACKAGES = System.getProperty("org.apache.avro.SERIALIZABLE_PACKAGES",
+        "java.lang,java.math,java.io,java.net,org.apache.avro.reflect").split(",");
+  }
+
+  private List<String> trustedPackages = new ArrayList<String>();
+  private boolean trustAllPackages = false;
```
[the remark this answers is not in the document]
Comment (jbonofre) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:45]: it was the initial idea, but it could be confusing for users/contributors. Let me improve a bit this part to be clearer.

**Comment by jbonofre** [comment]
@Fokko @martin-g @KalleOlaviNiemitalo I updated the PR. Can you guys please take a look ? Thanks !

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 5]
```diff
@@ -101,12 +115,43 @@ private Class getPropAsClass(Schema schema, String prop) {
...
+        }
+      }
+      if (!found) {
+        throw new ClassNotFoundException("Forbidden " + clazz
+            + "! This class is not trusted to be included in Avro schema using java-class. Please set org.apache.avro.SERIALIZABLE_PACKAGES system property with the packages you trust.");
+      }
+    }
+  }
+
+  public List<String> getTrustedPackages() {
```
Comment (martin-g) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:151]: This method should be `final` or it should be used at https://github.com/apache/avro/pull/2934/files#diff-95af304493048b22c3dda8fbfc7efb0e0c1c5e876f0234c2b8d0635bb2bcb494R127

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 6]
```diff
@@ -24,12 +24,25 @@
...
 public class SpecificDatumReader<T> extends GenericDatumReader<T> {
+
+  public static final String[] SERIALIZABLE_PACKAGES;
+
+  static {
+    SERIALIZABLE_PACKAGES = System.getProperty("org.apache.avro.SERIALIZABLE_PACKAGES",
+        "java.lang,java.math,java.io,java.net,org.apache.avro.reflect").split(",");
+  }
+
+  private List<String> trustedPackages = new ArrayList<>();
```
Comment (martin-g) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:44]:
```suggestion
  private final List<String> trustedPackages = new ArrayList<>();
```

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 7]
```diff
@@ -101,12 +115,43 @@ private Class getPropAsClass(Schema schema, String prop) {
...
+      Class clazz = ClassUtils.forName(getData().getClassLoader(), name);
+      checkSecurity(clazz);
+      return clazz;
     } catch (ClassNotFoundException e) {
       throw new AvroRuntimeException(e);
     }
   }
 
+  private boolean trustAllPackages() {
+    return (trustedPackages.size() == 1 && trustedPackages.get(0).equals("*"));
```
Comment (martin-g) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:127]:
```suggestion
    return (trustedPackages.size() == 1 && "*".equals(trustedPackages.get(0)));
```

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 8]
```diff
@@ -101,12 +115,43 @@ private Class getPropAsClass(Schema schema, String prop) {
...
+    if (thePackage != null) {
+      for (String trustedPackage : getTrustedPackages()) {
+        if (thePackage.getName().equals(trustedPackage) || thePackage.getName().startsWith(trustedPackage + ".")) {
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        throw new ClassNotFoundException("Forbidden " + clazz
+            + "! This class is not trusted to be included in Avro schema using java-class. Please set org.apache.avro.SERIALIZABLE_PACKAGES system property with the packages you trust.");
```
Comment (martin-g) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:146]: IMO the exception should mention the problematic class name. Currently it is hard for the developer to realize which class causes the problem.

Comment (martin-g) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:146]: Oh, right!
False alarm!

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 9]
```diff
@@ -24,12 +24,25 @@
 import org.apache.avro.io.ResolvingDecoder;
 import org.apache.avro.util.ClassUtils;
 import java.io.IOException;
+import java.util.ArrayList;
```
Comment (Fokko) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:27]:
```suggestion
import java.lang.SecurityException;
import java.util.ArrayList;
```

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java (hunk 5 above)
Comment (Fokko) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:151]:
```suggestion
  public final List<String> getTrustedPackages() {
```

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 10]
```diff
@@ -101,12 +115,43 @@ private Class getPropAsClass(Schema schema, String prop) {
...
+    Package thePackage = clazz.getPackage();
+    if (thePackage != null) {
+      for (String trustedPackage : getTrustedPackages()) {
+        if (thePackage.getName().equals(trustedPackage) || thePackage.getName().startsWith(trustedPackage + ".")) {
+          found = true;
+          break;
+        }
+      }
+      if (!found) {
+        throw new ClassNotFoundException("Forbidden " + clazz
```
[the remark this answers is not in the document]
Comment (Fokko) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:146]:
```suggestion
        throw new SecurityException("Forbidden " + clazz
```

[the remark this answers is not in the document]
Comment (jbonofre) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java]: Good point. I think you are right, it makes sense to throw `SecurityException` here. I will do that.

**Review by martin-g — APPROVED** [review]

**Comment by Fokko** [comment]
Moving this forward, thanks @jbonofre for fixing this 🙌  and thanks @martin-g and @KalleOlaviNiemitalo for the reviews!

File: lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java [hunk 11]
```diff
@@ -101,12 +115,43 @@ private Class getPropAsClass(Schema schema, String prop) {
...
+  }
+
+  private void checkSecurity(Class clazz) throws ClassNotFoundException {
+    if (trustAllPackages() || clazz.isPrimitive()) {
+      return;
+    }
+
+    boolean found = false;
+    Package thePackage = clazz.getPackage();
+    if (thePackage != null) {
```
Comment (MichalFoksa) [inline lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:137]: @Fokko or @martin-g 
When a class is not in any package `clazz.getPackage()` returns null and this condition renders that class trusty.

Move `if (!found) throw new SecurityException` out of the loop.

Comment (jbonofre) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:137]: I can take a look on this too as I'm the original author

Comment (martin-g) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:137]: @MichalFoksa Would you like to send a Pull Request?

Comment (jbonofre) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:137]: Else I can work on a PR tomorrow

Comment (MichalFoksa) [reply lang/java/avro/src/main/java/org/apache/avro/specific/SpecificDatumReader.java:137]: @martin-g  Here you are PR #3311
